### DISCUSSION TITLE
Any help migrating from Solid 0.10 to 1.x?

### DISCUSSION THREAD

**Discussion by bluenote10** [solidjs/solid#734] [Q&A] [answered]
I'd like to migrate [tabloo](https://github.com/bluenote10/tabloo) to a modern Solid version. I haven't touched the code in a long time (it was merely a hacky proof of concept anyway), and so far it is based on Solid 0.10.3. Of course I have now missed lots of changes in Solid, so I'm a bit lost. Is there something like a migration guide from old versions?

I assume the key challenges for the migration are:

- Version 0.10 still uses the `{(...)}` expression mechanism (which is also the reason why the code was never `prettier` formatted, which I'd like to add now as well). I vaguely remember that this was dropped (iirc in https://github.com/solidjs/solid/issues/91), but I still don't fully understand what kind of migration pattern I should use to get 100% the same behavior?
- Tabloo requires some [dom lifecycle hooks](https://github.com/bluenote10/tabloo/blob/master/src_frontend/dom_lifecycle.ts) that @ryansolid had proposed in [this issue](https://github.com/solidjs/solid/issues/36) / [this codesandbox](https://codesandbox.io/s/dom-lifecycle-u0gbd?file=/dom-lifecycle.js). Does this have a new equivalent in modern Solid?
- Any other difference I'm missing?

Any help would be highly appreciated :wink:

**Comment by ryansolid** [comment, marked as answer]
Hey long time no see. These changes were gradual over time so no migration guide.
1. The heuristic as introduced in the RFC is still present. Everything with a call expression (`something()`) or a member expression(`something.something`) access is wrapped. So if you don't want something to be tracked you need to pass just a variable `something` or use `/*@once*/` annotation.
2. We never made Solid DOM aware but we added scheduling to the reactive system. `createEffect` now runs after other rendering effects and in so as long things are in normal flow (ie you attach it as part of rendering) it will run after DOM elements have been attached. We also have an `onMount` that is just a `createEffect` with an `untrack` inside so it only runs once. `onCleanup` still handles the reactive graph cleanup. I'm not sure if this is a perfect swap but it's what we generally do now.
3. Similarly it means that the execution timing of `createEffect` is slightly different. The old timing is found in `createComputed` primitive. However, in modern Solid this primitive is only used for synchronizing other signals. Generally most side effects that you'd see like ones that interact with the DOM are preferred to be in `createEffect`.
4. There was some naming changes I think from then `freeze` -> `batch`, `sample` -> `untrack`, `map` -> `mapArray`, `createState` -> `createStore`. `createStore` is found under `solid-js/store` now. 
5. `<For>` now has index as second argument as a signal. So no more need for like `<ForIndex>`.

Not sure what else.

**Comment by bluenote10** [comment]
Thanks for the summary, that has definitely helped a lot!

For the record, the things that were a bit of a challenge were:
- At first I had many issues with getting the code to compile JSX expressions again. I turned out to be related to the change mentioned [here](https://github.com/solidjs/solid/issues/282#issuecomment-747203868). What worked in the end was to
    
    - remove `babel-plugin-jsx-dom-expressions` and add `babel-preset-solid` in `package.json`.
    - add `"jsxImportSource": "solid-js"` and remove the `"./node_modules/babel-plugin-jsx-dom-expressions"` from the `typeRoots` in `tsconfig.json`.
    - introduce imports of `JSX` from `solid-js`.

- The other issue was related to mount/unmount. As mentioned in the original issue, I mainly opted for Solid over React, because it allows to "prepare" the dom elements in the background. This makes switching the tabs super fast, because I don't have to rebuild any complex dom structures or plots every time the tab changes, but I can keep them prepared, and just re-attach / detach them from the dom. This can be particularly nice with WebGL based libraries that need to re-buffer vertex data every time the plot enters the dom. In any case, `createEffect` and `onMount` cannot be used in this case, because `onMount` isn't really an "on mount", but fires as soon the element gets created in the background. However, I need to delay certain (canvas-requiring) calls until the element really enters the dom (in this example echart's zrender will spam with exceptions when trying to render into a non-visible element, which is often the case with plotting libraries). Fortunately, I figured out that I could make the mutation observer based approach work again to get true mount/unmount callbacks.
